### PR TITLE
Chapter 5 Exercise 4

### PR DESCRIPTION
**Pull request by yehudamber**
Add a command-line option to also print the content of subdirectories.

### CODE REVIEW

File: ch5-catdir/controller.cpp [hunk 1]
```diff
@@ -26,13 +27,31 @@ Controller::~Controller()
...
+
+    if (m_args.m_showHelp)
+    {
+        m_err << "Usage: " << m_args.m_progName << " [-hr] [DIRECTORY]...\n"
+                 "Print the content of the DIRECTORY(ies).\n\n"
+                 "With no DIRECTORY, use the current working directory.\n"
+                 "Exit status is 0 for success, 1 for failure and 2 for"
+                 " command-line error.\n\n"
+                 "  -h\tShow this help and exit\n"
+                 "  -r\tAlso print the content of subdirectories\n";
```
Comment (YehezkelShB) [inline ch5-catdir/controller.cpp:44]: More commonly phrased as: `recurse into directories` (from `zip` and `tar` help)

File: ch5-catdir/controller.cpp [hunk 2]
```diff
@@ -55,6 +74,10 @@ void Controller::process(const std::string_view& path) try
     auto dir = Directory(path);
     while (auto ent = dir.next())
     {
+        if (!isArgument)
+        {
+            m_out << '\t';
+        }
```
Comment (YehezkelShB) [inline ch5-catdir/controller.cpp:80]: IIUC, you add a single tab before the info, only for subdirectories not given as an argument in the command-line.
What about subdir of a subdir? Do you want them to be nested further (with 2 tabs) or just 1 tab for all of them?

File: ch5-catdir/controller.cpp [hunk 3]
```diff
@@ -42,11 +61,11 @@ int Controller::run()
...
 std::ostream& Controller::error()
 {
-    m_instance->m_exitStatus = EXIT_FAILURE;
-    return m_instance->m_err << m_instance->m_progName << ": ";
+    m_instance->m_exitStatus = ExitStatus::FAILURE;
+    return m_instance->m_err << m_instance->m_args.m_progName << ": ";
 }
 
-void Controller::process(const std::string_view& path) try
+void Controller::process(const std::string_view& path, bool isArgument) try
```
Comment (YehezkelShB) [inline ch5-catdir/controller.cpp]: `isArgument` sounds a bit strange to me. Everything is an argument when passed to a function... :)
Maybe `isCmdLineArgument`? I'm not sure myself

File: ch5-catdir/controller.h [hunk 4]
```diff
@@ -1,10 +1,22 @@
...
-#include <cstdlib>
 #include <iosfwd>
 #include <span>
 #include <string_view>
 
+/*
+ * ExitStatus - constants for the exit status values returned by the program
+ */
+
+namespace ExitStatus
```
Comment (YehezkelShB) [inline ch5-catdir/controller.h:11]: Not `enum class`, to avoid casting to `int`? Interesting.

**Review by YehezkelShB — APPROVED** [review]

File: ch5-catdir/controller.h (hunk 4 above)
Comment (yehudamber) [reply ch5-catdir/controller.h:11]: But I _want_ it to be `int`, which is the type used for the exit status...

File: ch5-catdir/controller.cpp (hunk 2 above)
Comment (yehudamber) [reply ch5-catdir/controller.cpp:80]: Note that I don't read subdirs of subdirs at all:
if (**isArgument** && m_args.m_readSubdirs && ent->isSubdir()

File: ch5-catdir/controller.cpp (hunk 1 above)
Comment (yehudamber) [reply ch5-catdir/controller.cpp:44]: See my comment on the next conversation.

**Review by YehezkelShB — APPROVED** [review]

**Pull request merged**